test/java/integration/AndroidEngineSocketTagTest.java

```diff
@@ -0,0 +1,392 @@
+    final Iterator<ByteBuffer> chunkIterator = requestScenario.getBodyChunks().iterator();
+
+    Stream stream =
+        engine.streamClient()
+            .newStreamPrototype()
+            .setOnSendWindowAvailable(ignored -> {
+              onSendWindowAvailable(requestScenario, streamRef.get(), chunkIterator,
+                                    response.get());
+              return null;
+            })
+            .setOnResponseHeaders((responseHeaders, endStream, streamIntel) -> {
+              response.get().setHeaders(responseHeaders);
+              response.get().addStreamIntel(streamIntel);
+              if (requestScenario.cancelOnResponseHeaders) {
+                streamRef.get().cancel(); // Should be a noop when endStream == true
+              } else {
+                if (requestScenario.waitOnReadData) {
+                  try {
+                    Thread.sleep(100 + (int)(Math.random() * 50));
+                  } catch (InterruptedException e) {
+                    // Don't care
+                  }
+                }
+                streamRef.get().readData(requestScenario.responseBufferSize);
+              }
+              return null;
+            })
+            .setOnResponseData((data, endStream, streamIntel) -> {
+              response.get().addBody(data);
+              response.get().addStreamIntel(streamIntel);
+              if (!endStream) {
+                if (requestScenario.waitOnReadData) {
+                  try {
+                    Thread.sleep(100 + (int)(Math.random() * 50));
+                  } catch (InterruptedException e) {
+                    // Don't care
+                  }
+                }
+                streamRef.get().readData(requestScenario.responseBufferSize);
+              }
+              return null;
+            })
+            .setOnResponseTrailers((trailers, streamIntel) -> {
+              response.get().setTrailers(trailers);
+              response.get().addStreamIntel(streamIntel);
+              return null;
+            })
+            .setOnError((error, finalStreamIntel) -> {
+              response.get().setEnvoyError(error);
+              response.get().addStreamIntel(finalStreamIntel);
+              response.get().setFinalStreamIntel(finalStreamIntel);
+              latch.countDown();
+              return null;
+            })
+            .setOnCancel((finalStreamIntel) -> {
+              response.get().setCancelled();
+              response.get().addStreamIntel(finalStreamIntel);
+              response.get().setFinalStreamIntel(finalStreamIntel);
+              latch.countDown();
+              return null;
+            })
+            .setOnComplete((finalStreamIntel) -> {
+              response.get().addStreamIntel(finalStreamIntel);
+              response.get().setFinalStreamIntel(finalStreamIntel);
+              latch.countDown();
+              return null;
+            })
+            .setExplicitFlowControl(true)
```

Comment:
just thinking (and skip if it takes too long) I wonder if you kill off explicit flow control you can remove the window update functions.  and if you're not using intel you can probably remove most of these callback functions.  and if not just ignore me.